From 8a9985fcc931c1bc51755330e9ae721f8d31cb83 Mon Sep 17 00:00:00 2001
From: Marsell Kukuljevic <marsell@joyent.com>
Date: Mon, 24 Sep 2018 21:08:17 +1200
Subject: [PATCH] IMAGE-1135 Need a memo attribute in manifest for Triton to
 tell if the image can be provisioned as bhyve VM

---
 .gitignore          |  1 +
 CHANGES.md          |  3 ++-
 lib/imgmanifest.js  | 51 ++++++++++++++++++++++++++++++++++---
 test/basics.test.js | 62 +++++++++++++++++++++++++++++++++++++++++++++
 4 files changed, 112 insertions(+), 5 deletions(-)

diff --git a/.gitignore b/.gitignore
index 33f4884..845e9ec 100644
--- a/.gitignore
+++ b/.gitignore
@@ -4,3 +4,4 @@ build
 cscope.in.out
 cscope.po.out
 cscope.out
+*.swp
diff --git a/CHANGES.md b/CHANGES.md
index b99ac00..b444ae3 100644
--- a/CHANGES.md
+++ b/CHANGES.md
@@ -2,7 +2,8 @@
 
 ## not yet released
 
-(nothing yet)
+- IMAGE-1135 Need a memo attribute in manifest for Triton to tell if the image
+  can be provisioned as bhyve VM
 
 ## 3.1.0
 
diff --git a/lib/imgmanifest.js b/lib/imgmanifest.js
index bfdc99e..54d2135 100644
--- a/lib/imgmanifest.js
+++ b/lib/imgmanifest.js
@@ -1265,6 +1265,15 @@ var validators = {
         }
         delete reqs.networks;
 
+        if (reqs.brand && reqs.brands) {
+            errs.push({
+                field: 'requirements.brand',
+                code: 'Invalid',
+                message: 'Only one of "requirements.brand" and ' +
+                    '"requirements.brands" can be present'
+            });
+        }
+
         // requirements.brand
         if (reqs.brand && typeof (reqs.brand) !== 'string') {
             errs.push({
@@ -1277,6 +1286,32 @@ var validators = {
         }
         delete reqs.brand;
 
+        // requirements.brands
+        if (reqs.brands) {
+            if (Array.isArray(reqs.brands)) {
+                reqs.brands.filter(function checkBrandType(brand) {
+                    return typeof (brand) !== 'string';
+                }).forEach(function addErr(brand) {
+                    errs.push({
+                        field: 'requirements.brands',
+                        code: 'Invalid',
+                        message: format(
+                        'invalid image "%s" (not a string): %j',
+                        'requirements.brands', brand)
+                    });
+                });
+            } else {
+                errs.push({
+                    field: 'requirements.brands',
+                    code: 'Invalid',
+                    message: format(
+                    'invalid image "requirements.brands" (not an array): %j',
+                    reqs.brands)
+                });
+            }
+        }
+        delete reqs.brands;
+
         // requirements.ssh_key
         if (reqs.ssh_key && typeof (reqs.ssh_key) !== 'boolean') {
             errs.push({
@@ -1386,8 +1421,15 @@ var validators = {
 
         if (reqs.bootrom) {
             var VALID_BOOTROMS = ['bios', 'uefi'];
-            var brand = manifest.requirements.brand;
-            if (!brand) {
+
+            // due to checks above, only requirements.brand or
+            // requirements.brands can be present at this point
+            var brands = manifest.requirements.brands || [];
+            if (manifest.requirements.brand) {
+                brands.push(manifest.requirements.brand);
+            }
+
+            if (brands.length === 0) {
                 errs.push({
                     field: 'requirements.bootrom',
                     code: 'MissingParameter',
@@ -1395,13 +1437,14 @@ var validators = {
                         + '"requirements.bootrom" requires '
                         + '"requirements.brand" to be set to "bhyve"'
                 });
-            } else if (brand !== 'bhyve') {
+            } else if (brands.indexOf('bhyve') === -1) {
                 errs.push({
                     field: 'requirements.bootrom',
                     code: 'Invalid',
                     message: format('invalid bootrom '
                         + '"requirements.bootrom" not supported with "%s" '
-                        + 'brand; "requirements.brand" must be "bhyve"', brand)
+                        + 'brand(s); "requirements.brands" must include '
+                        + '"bhyve"', brands.join())
                 });
             } else if (VALID_BOOTROMS.indexOf(reqs.bootrom) === -1) {
                 errs.push({
diff --git a/test/basics.test.js b/test/basics.test.js
index 2755b1c..46c2227 100644
--- a/test/basics.test.js
+++ b/test/basics.test.js
@@ -227,6 +227,45 @@ var MINIMAL_VALIDATIONS = [
             }
         }
     },
+    {
+        name: 'good minimal, requirements.brands, requirements.bootrom',
+        manifest: {
+            'v': 2,
+            'uuid': '1f9b7958-289e-4ea3-8f88-5486a40d6823',
+            'urn': 'booga:booga',
+            'name': 'foo',
+            'version': '1.2.3',
+            'type': 'zvol',
+            'os': 'linux',
+            'requirements': {
+                'brands': ['bhyve', 'kvm'],
+                'bootrom': 'bios'
+            }
+        }
+    },
+    {
+        name: 'bad minimal, requirements.brand and requirements.brands',
+        errs: [
+            {
+                field: 'requirements.brand',
+                code: 'Invalid',
+                message: /one of .requirements.brand. and .requirements.brands./
+            }
+        ],
+        manifest: {
+            'v': 2,
+            'uuid': '1f9b7958-289e-4ea3-8f88-5486a40d6823',
+            'urn': 'booga:booga',
+            'name': 'foo',
+            'version': '1.2.3',
+            'type': 'zvol',
+            'os': 'linux',
+            'requirements': {
+                'brand': 'bhyve',
+                'brands': []
+            }
+        }
+    },
     {
         name: 'bad minimal, requirements.bootrom=/some/path',
         errs: [
@@ -294,6 +333,29 @@ var MINIMAL_VALIDATIONS = [
                 'bootrom': 'uefi'
             }
         }
+    },
+    {
+        name: 'bad minimal, requirements.bootrom with requirements.brands=kvm',
+        errs: [
+            {
+                field: 'requirements.bootrom',
+                code: 'Invalid',
+                message: /not supported with .kvm. brand/
+            }
+        ],
+        manifest: {
+            'v': 2,
+            'uuid': '1f9b7958-289e-4ea3-8f88-5486a40d6823',
+            'urn': 'booga:booga',
+            'name': 'foo',
+            'version': '1.2.3',
+            'type': 'zvol',
+            'os': 'linux',
+            'requirements': {
+                'brands': ['kvm'],
+                'bootrom': 'uefi'
+            }
+        }
     }
 ];
 
-- 
2.21.0

